### ISSUE TITLE
"Illegal instruction" when compiling program using library

### ISSUE THREAD

**Issue by perdugames** [Auburns/FastNoiseSIMD#20]
I still can not get it to work, I'm now experiencing a flaw in segmentation when trying to start a new class in my module. I get this error while running my program: https://pastebin.com/V6tMFJ3F

I have this pointer:

`FastNoiseSIMD* myNoise;`

the builder of my class:

```
GDFastNoiseSIMD::GDFastNoiseSIMD(){

    myNoise = FastNoiseSIMD::NewFastNoiseSIMD();

}
```

and thus I call the methods:

`myNoise->GetSeed();`

Am I doing something wrong? 

The error is accusing these lines below:

https://github.com/Auburns/FastNoiseSIMD/blob/master/FastNoiseSIMD/FastNoiseSIMD.h#L98

https://github.com/Auburns/FastNoiseSIMD/blob/master/FastNoiseSIMD/FastNoiseSIMD.cpp#L207

https://github.com/Auburns/FastNoiseSIMD/blob/master/FastNoiseSIMD/FastNoiseSIMD_internal.cpp#L1028

I also created an empty program to test and confirm the error "Illegal instruction", probably accusing the segmentation failure. Follow the zip with the program and everything needed to reproduce the problem.

[test_fast_simd.zip](https://github.com/Auburns/FastNoiseSIMD/files/1569240/test_fast_simd.zip)

**Comment by Auburn** [comment]
Make sure you having everything setup correctly from here: https://github.com/Auburns/FastNoiseSIMD/wiki/Compiler-Settings

**Comment by perdugames** [comment]
I'm using the flags. The output of scons when compiling:

**Thread model: posix
gcc version 7.2.0 (GCC)**

```
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
g++ -o FastNoiseSIMD/FastNoiseSIMD/FastNoiseSIMD.o -c -std=c++11 -IFastNoiseSIMD/FastNoiseSIMD -I. FastNoiseSIMD/FastNoiseSIMD/FastNoiseSIMD.cpp
g++ -o FastNoiseSIMD/FastNoiseSIMD/FastNoiseSIMD_avx2.o -c -std=c++11 -march=core-avx2 -IFastNoiseSIMD/FastNoiseSIMD FastNoiseSIMD/FastNoiseSIMD/FastNoiseSIMD_avx2.cpp
g++ -o FastNoiseSIMD/FastNoiseSIMD/FastNoiseSIMD_avx512.o -c -std=c++11 -IFastNoiseSIMD/FastNoiseSIMD -I. FastNoiseSIMD/FastNoiseSIMD/FastNoiseSIMD_avx512.cpp
g++ -o FastNoiseSIMD/FastNoiseSIMD/FastNoiseSIMD_internal.o -c -std=c++11 -IFastNoiseSIMD/FastNoiseSIMD -I. FastNoiseSIMD/FastNoiseSIMD/FastNoiseSIMD_internal.cpp
g++ -o FastNoiseSIMD/FastNoiseSIMD/FastNoiseSIMD_neon.o -c -std=c++11 -IFastNoiseSIMD/FastNoiseSIMD -I. FastNoiseSIMD/FastNoiseSIMD/FastNoiseSIMD_neon.cpp
g++ -o FastNoiseSIMD/FastNoiseSIMD/FastNoiseSIMD_sse2.o -c -std=c++11 -msse2 -IFastNoiseSIMD/FastNoiseSIMD FastNoiseSIMD/FastNoiseSIMD/FastNoiseSIMD_sse2.cpp
g++ -o FastNoiseSIMD/FastNoiseSIMD/FastNoiseSIMD_sse41.o -c -std=c++11 -msse4.1 -IFastNoiseSIMD/FastNoiseSIMD FastNoiseSIMD/FastNoiseSIMD/FastNoiseSIMD_sse41.cpp
g++ -o test.o -c -std=c++11 -IFastNoiseSIMD/FastNoiseSIMD -I. test.cpp
g++ -o test test.o FastNoiseSIMD/FastNoiseSIMD/FastNoiseSIMD_avx512.o FastNoiseSIMD/FastNoiseSIMD/FastNoiseSIMD_internal.o FastNoiseSIMD/FastNoiseSIMD/FastNoiseSIMD_neon.o FastNoiseSIMD/FastNoiseSIMD/FastNoiseSIMD.o FastNoiseSIMD/FastNoiseSIMD/FastNoiseSIMD_avx2.o FastNoiseSIMD/FastNoiseSIMD/FastNoiseSIMD_sse41.o FastNoiseSIMD/FastNoiseSIMD/FastNoiseSIMD_sse2.o
scons: done building targets.
```

**Comment by perdugames** [comment]
I commented the "**#define FN_COMPILE_AVX2**" in _FastNoiseSIMD.h_ and it worked now. But here https://github.com/Auburns/FastNoiseSIMD/wiki/Compiler-Settings you said that "G ++ / GCC 4.7+" is supported, so I had not even tried to comment, I thought my compiler supported, but had forgotten to check if my CPU supported, my i3 does not support avx2. Now I still have the doubt, if I commented the "**#define FN_COMPILE_AVX2**" what does it do? the computers that require these instructions will not be able to run with the parallelization and will have to run with the single-thread mode why I did not compile with "**#define FN_COMPILE_AVX2**"?

**Comment by PerduGames** [comment]
I am not understanding the values I get:

```
#include "FastNoiseSIMD.h"
#include <stdio.h>

int main()
{

    FastNoiseSIMD* myNoise = FastNoiseSIMD::NewFastNoiseSIMD();

    myNoise->SetNoiseType(FastNoiseSIMD::Simplex);

    float* noiseSet = myNoise->GetNoiseSet(0, 0, 0, 1, 2, 2);

    int index = 0;
    for (int x = 0; x < 1; x++)
    {
        for (int y = 0; y < 2; y++)
        {
            for (int z = 0; z < 2; z++)
            {
                printf ("n:%d \n", noiseSet[index++]);
            }
        }
    }

    FastNoiseSIMD::FreeNoiseSet(noiseSet);
    return 0;
}
```
Output from printf:

```
n:38298368 
n:0 
n:0 
n:0
```

**Comment by Auburn** [comment]
Hmm, it looks like the same issue that happens when you have "Whole program optimisation" turned on in MSVC. Which causes SIMD code to get inlined where it shouldn't which causes the illegal instruction crash.

Try changing FastNoiseSIMD_internal.cpp#L609 from:
`void FUNC(InitSIMDValues)()`
to
`void __attribute__ ((noinline)) FUNC(InitSIMDValues)()`

and uncomment "#define FN_COMPILE_AVX2"

**Comment by perdugames** [comment]
I tried to make the change below:

```
Try changing FastNoiseSIMD_internal.cpp#L609 from:
void FUNC(InitSIMDValues)()
to
void __attribute__ ((noinline)) FUNC(InitSIMDValues)()
```

but I still receive "**illegal instruction**". And I am not activating any optimization.


When I comment "#define FN_COMPILE_AVX2" I can compile, and try the code below:

```
#include "FastNoiseSIMD.h"
#include <stdio.h>

int main()
{

    FastNoiseSIMD* myNoise = FastNoiseSIMD::NewFastNoiseSIMD(26546);

    myNoise->SetNoiseType(FastNoiseSIMD::Simplex);
    float* noiseSet = myNoise->GetNoiseSet(0, 0, 0, 2, 2, 2);

    int index = 0;
    for (int x = 0; x < 2; x++)
    {
        for (int y = 0; y < 2; y++)
        {
            for (int z = 0; z < 2; z++)
            {
                printf ("n:%f \n", noiseSet[index++]); // or *(noiseSet + index++)
            }
        }
    }

    FastNoiseSIMD::FreeNoiseSet(noiseSet);
    return 0;
}

```
I get:

```
n:0.000000 
n:0.041444 
n:-0.041444 
n:0.000000 
n:0.000000 
n:0.041417 
n:-0.041417 
n:0.000000 

```
There are many "0", is this normal?

**Comment by Auburn** [comment]
I'm not sure what's going wrong, are you able to test with MSVC to see if
you have the same issues, it has had more testing than GCC/G++

**Comment by perdugames** [comment]
I do not have MSVC. I use voidlinux. not sure about "illegal instruction" or "0"?

And what I'm working on. https://github.com/PerduGames/gdfastnoisesimd

when I end this module for my Godot client, I intend to create a module for nodejs as well, because that's where my server is. I need FastNoiseSIMD for both client and server.

Edit: I realize that I only get a lot of "0" so with **Simplex** and **SimplexFractal** noise. And I've already started working on the module for nodejs. The Godot module is practically ready, all we need is an understanding of these "0's" and the case of "avx2".

**Comment by perdugames** [comment]
Could not you create a shared library and static already with everything compiled? Would not that work?

**Comment by Auburn** [comment]
I'm not sure what is causing your issues, if it helps here is the StarMade native code jdk plugin used for all platforms. https://github.com/Schine/StarMadeNative

**Comment by perdugames** [comment]
But what about the "0's", is it for me to receive them myself, or is something wrong?

**Comment by Auburn** [comment]
Could you try a noise set of size 8, 8, 8 and see if you still get 0s

**Comment by perdugames** [comment]
I'm getting "0" for every 8 elements. This is a pattern, something is going wrong.

**Simplex 8x8x8**
```
myNoise->SetNoiseType(FastNoiseSIMD::Simplex);
float* noiseSet = myNoise->GetNoiseSet(0, 0, 0, 8, 8, 8);
```
```
n:0.000000 
n:0.041444 
n:0.082723 
n:0.123671 
n:0.164126 
n:0.203926 
n:0.242914 
n:0.280936 
n:-0.041444 
n:0.000000 
n:0.041334 
n:0.082392 
n:0.123012 
n:0.163031 
n:0.202292 
n:0.240641 
n:-0.082723 
n:-0.041334 
n:0.000000 
n:0.041114 
n:0.081844 
n:0.122028 
n:0.161508 
n:0.200130 
n:-0.123671 
n:-0.082392 
n:-0.041114 
n:0.000000 
n:0.040785 
n:0.081080 
n:0.120725 
n:0.159566 
n:-0.164126 
n:-0.123012 
n:-0.081844 
n:-0.040785 
n:0.000000 
n:0.040350 
n:0.080106 
n:0.119112 
n:-0.203926 
n:-0.163031 
n:-0.122028 
n:-0.081080 
n:-0.040350 
n:0.000000 
n:0.039811 
n:0.078927 
n:-0.242914 
n:-0.202292 
n:-0.161508 
n:-0.120725 
n:-0.080106 
n:-0.039811 
n:0.000000 
n:0.039171 
n:-0.280936 
n:-0.240641 
n:-0.200130 
n:-0.159566 
n:-0.119112 
n:-0.078927 
n:-0.039171 
n:0.000000 
n:0.000000 
n:0.041417 
n:0.082668 
n:0.123589 
n:0.164016 
n:0.203789 
n:0.242751 
n:0.280748 
n:-0.041417 
n:0.000000 
n:0.041306 
n:0.082337 
n:0.122930 
n:0.162922 
n:0.202157 
n:0.240479 
n:-0.082668 
n:-0.041306 
n:0.000000 
n:0.041086 
n:0.081789 
n:0.121946 
n:0.161400 
n:0.199995 
n:-0.123589 
n:-0.082337 
n:-0.041086 
n:0.000000 
n:0.040758 
n:0.081026 
n:0.120644 
n:0.159459 
n:-0.164016 
n:-0.122930 
n:-0.081789 
n:-0.040758 
n:0.000000 
n:0.040323 
n:0.080052 
n:0.119031 
n:-0.203789 
n:-0.162922 
n:-0.121946 
n:-0.081026 
n:-0.040323 
n:0.000000 
n:0.039784 
n:0.078874 
n:-0.242751 
n:-0.202157 
n:-0.161400 
n:-0.120644 
n:-0.080052 
n:-0.039784 
n:0.000000 
n:0.039145 
n:-0.280748 
n:-0.240479 
n:-0.199995 
n:-0.159459 
n:-0.119031 
n:-0.078874 
n:-0.039145 
n:0.000000 
n:0.000000 
n:0.041334 
n:0.082503 
n:0.123341 
n:0.163687 
n:0.203380 
n:0.242263 
n:0.280182 
n:-0.041334 
n:0.000000 
n:0.041224 
n:0.082173 
n:0.122683 
n:0.162595 
n:0.201750 
n:0.239994 
n:-0.082503 
n:-0.041224 
n:0.000000 
n:0.041004 
n:0.081625 
n:0.121701 
n:0.161075 
n:0.199592 
n:-0.123341 
n:-0.082173 
n:-0.041004 
n:0.000000 
n:0.040676 
n:0.080863 
n:0.120401 
n:0.159137 
n:-0.163687 
n:-0.122683 
n:-0.081625 
n:-0.040676 
n:0.000000 
n:0.040242 
n:0.079891 
n:0.118791 
n:-0.203380 
n:-0.162595 
n:-0.121701 
n:-0.080863 
n:-0.040242 
n:0.000000 
n:0.039704 
n:0.078714 
n:-0.242263 
n:-0.201750 
n:-0.161075 
n:-0.120401 
n:-0.079891 
n:-0.039704 
n:0.000000 
n:0.039066 
n:-0.280182 
n:-0.239994 
n:-0.199592 
n:-0.159137 
n:-0.118791 
n:-0.078714 
n:-0.039066 
n:0.000000 
n:0.000000 
n:0.041196 
n:0.082228 
n:0.122930 
n:0.163140 
n:0.202700 
n:0.241451 
n:0.279241 
n:-0.041196 
n:0.000000 
n:0.041086 
n:0.081898 
n:0.122273 
n:0.162051 
n:0.201074 
n:0.239188 
n:-0.082228 
n:-0.041086 
n:0.000000 
n:0.040867 
n:0.081352 
n:0.121294 
n:0.160535 
n:0.198921 
n:-0.122930 
n:-0.081898 
n:-0.040867 
n:0.000000 
n:0.040540 
n:0.080592 
n:0.119997 
n:0.158602 
n:-0.163140 
n:-0.122273 
n:-0.081352 
n:-0.040540 
n:0.000000 
n:0.040107 
n:0.079622 
n:0.118391 
n:-0.202700 
n:-0.162051 
n:-0.121294 
n:-0.080592 
n:-0.040107 
n:0.000000 
n:0.039570 
n:0.078449 
n:-0.241451 
n:-0.201074 
n:-0.160535 
n:-0.119997 
n:-0.079622 
n:-0.039570 
n:0.000000 
n:0.038934 
n:-0.279241 
n:-0.239188 
n:-0.198921 
n:-0.158602 
n:-0.118391 
n:-0.078449 
n:-0.038934 
n:0.000000 

```

**Comment by perdugames** [comment]
What does "GetSampledNoiseSet" do? Sorry for the many questions.

**Comment by perdugames** [comment]
As for the "illegal instruction" problem that occurs when trying to run the program compiled with "#define FN_COMPILE_AVX2", I'm not sure why it occurs, my "SIMDLevel" is "2", ie "SSE41" so it calls this line: https://github.com/Auburns/FastNoiseSIMD/blob/master/FastNoiseSIMD/FastNoiseSIMD.cpp#L207 
and I get "illegal instruction", if it comments, it goes to the "SSE2" if on the line:
https://github.com/Auburns/FastNoiseSIMD/blob/master/FastNoiseSIMD/FastNoiseSIMD.cpp#L214
, I also receive "illegal instruction", commenting on it, I go to "FN_COMPILE_NO_SIMD_FALLBACK", on the line: https://github.com/Auburns/FastNoiseSIMD/blob/master/FastNoiseSIMD/FastNoiseSIMD.cpp#L218
, I no longer receive "illegal instruction" here. 
So I think something goes wrong with those macros in "FastNoiseSIMD_internal.cpp", the strange thing is that when I comment on "#define FN_COMPILE_AVX2", it works normally, despite the "0" pattern in the Simplex function and SimplexFractal. Then maybe there is something wrong going on in the macros that is causing these confusions to happen.

or gcc is not recognizing one instruction, and is not giving me any warning about it and causing it to fail. But I also tested with clang and I get "illegal instruction". And if I try to compile only sse2 when running the program I get a "segmentation fault"

**Comment by Auburn** [comment]
Try compile with all optimisations turned off, I feel like it is an inlining issue somewhere.

**Comment by perdugames** [comment]
I'm not compiling with any optimization. See the commands below, this causes me a "segmentation fault". Even with "#define FN_COMPILE_AVX2", "define FN_COMPILE_AVX512", "#define FN_COMPILE_SSE41" and "#define FN_ALIGNED_SETS" commented.

```
g++ -o FastNoiseSIMD/FastNoiseSIMD.o -c -std=c++11 -IFastNoiseSIMD -I. FastNoiseSIMD/FastNoiseSIMD.cpp
g++ -o FastNoiseSIMD/FastNoiseSIMD_internal.o -c -std=c++11 -IFastNoiseSIMD -I. FastNoiseSIMD/FastNoiseSIMD_internal.cpp
g++ -o FastNoiseSIMD/FastNoiseSIMD_sse2.o -c -std=c++11 -msse2 FastNoiseSIMD/FastNoiseSIMD_sse2.cpp
g++ -o test.o -c -std=c++11 -IFastNoiseSIMD -I. test.cpp
g++ -o test test.o FastNoiseSIMD/FastNoiseSIMD_internal.o FastNoiseSIMD/FastNoiseSIMD.o FastNoiseSIMD/FastNoiseSIMD_sse2.o
```

**Comment by perdugames** [comment]
And I analyzed this https://jordanpeck.me/2016/09/performance-comparisons/? would FastNoise2D be faster even simlulating only 2 dimensions and setting to "1" one of the axes in FastNoiseSIMD?

**Comment by Auburn** [comment]
The 0s in the Simplex algorithm is normal, its mainly because you are getting noise from the origin. Can you show the disassembly code where the illegal instruction happens?

**Comment by perdugames** [comment]
would be normal if it does not follow standard, but if there is a default, it will render the noise function unusable. And it's being a standard, every 8 elements I get a 0 like in the above example. I still do not know where this "illegal instruction" comes from. It is difficult to find its origin.

**Comment by caseymcc** [comment]
Coming a little late to this discussion but I ran into the same thing. I found out through stack overflow (https://stackoverflow.com/questions/52745772/compiling-library-with-sse2-and-avx2/52746875#52746875) that when you compile with /arch:AVX2 the sse2/4.1 instructions will use the vex prefix encoding causing them to fail on a machine without avx2 support. The different simd levels need to be compiled into separate shared libs with the different compile options. I have done this as well as converted the use of macros to templates here (https://github.com/caseymcc/HastyNoise)
